src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -476,6 +483,42 @@ internal bool RemoveParsedValue(HeaderDescriptor descriptor, object value)
             return false;
         }
 
+        private void RemoveParsedValue(HeaderStoreItemInfo info)
+        {
+            var result = new List<HeaderStoreItemInfoValue>();
+            foreach (var item in info.ParsedAndInvalidValues)
+            {
+                if (item.Type != HeaderStoreItemInfoValueType.Parsed)
+                {
+                    result.Add(item);
+                }
+            }
+
+            info.ParsedAndInvalidValues = result;
+        }
+
+        private static object? GetParsedValues(HeaderStoreItemInfo info)
```

Comment:
We should avoid allocating a new `List` here. The callers should be responsible for dealing with the `ParsedAndInvalidValues` field as needed without requiring this helper.

Reply:
@MihaZupan

if use `internal object ParsedAndInvalidValues;`, how to return original order of all values or parsed values?

ParsedValues and InvalidValues can be list also.

Reply:
The value of `ParsedAndInvalidValues` can still be a list, the difference being that it would start out as a single object, and the list would only be allocated if there is more than 1 value.

Reply:
@MihaZupan 
if ParsedAndInvalidValues  stored values ["Parsed1","Invalid1","Parsed2","Invalid2"],

how to get only parsed vlaues?

Reply:
Similar to https://github.com/dotnet/runtime/pull/65249#issuecomment-1040085563

```c#
if (ParsedAndInvalidValues is not null)
{
    if (ParsedAndInvalidValues is List<object> list)
    {
        foreach (object value in list)
        {
            if (value is not InvalidValue)
            {
                // "value" is a parsed value
            }
        }
    }
    else if (ParsedAndInvalidValues is not InvalidValue)
    {
        // ParsedAndInvalidValues is a single parsed value
    }
}
```

If it makes the code any less complex, we could choose to always allocate the list if there is an invalid value. But we should always avoid allocating one if all we are dealing with is a singe valid value.

Reply:
Note that maintaining such order for all headers would involve an even larger change.
The `CacheControlHeaderParser` for example will store multiple values in a single object, so you would lose the valid/invalid ordering there.